### PR TITLE
Add typescript support as an option

### PR DESCRIPTION
**Pull request by BPScott**
Sewing-kit recently enabled support for babel reading typescript files directly when running jest tests, it did that by defining its own config instead of extending babel-preset-shopify. To help avoid repetition and to suggest a canonical way to support typescript I think we should build the required config into babel-preset-shopify.

This PR adds a `typescript` option to the web and node presets. This defaults to false, which causes no changes. When set to true: `@babel/preset-typescript` is added to the list of presets used and `@babel/plugin-proposal-decorators` and `@babel/plugin-proposal-numeric-separator` are added to the plugins to enable support for typescript functionality.

This also updates the node preset to always enable the `babel-plugin-dynamic-import-node` plugin as from what I can tell by reading sewing-kit source code we always want to enable support for the `import()` syntax in node-land. @GoodForOneFare / @lemonmade have I interpreted that right or should I only enable this when in typescript mode?

---

As an example of usage this will let us replace [sewing-kit's babel.config.js](https://github.com/Shopify/sewing-kit/blob/b41d39e167c15e46258dae1a3555b53c09c27cea/babel.config.js) with:

```js
module.exports = {
  presets: [['babel-preset-shopify/node', {typescript: true}]],
  sourceMaps: 'inline',
};
```

### CODE REVIEW

File: non-standard-plugins.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 module.exports = function shopifyNonStandardPlugins(options = {}) {
-  const {inlineEnv = false} = options;
+  const {inlineEnv = false, typescript = false} = options;
 
   const plugins = [
-    require.resolve('@babel/plugin-proposal-class-properties'),
+    [require.resolve('@babel/plugin-proposal-class-properties'), {loose: typescript}],
```
Comment (lemonmade) [inline non-standard-plugins.js]: I originally asked why this was needed. Comment below explains it very well (thanks for that :)) but a bit weird that it occurs after. Maybe just unshift this and the decorators plugin in the if (typescript) below, then add an else to unshift only this one?

File: non-standard-plugins.js [hunk 2]
```diff
@@ -12,5 +12,21 @@ module.exports = function shopifyNonStandardPlugins(options = {}) {
...
+    // Typscript implements the stage 1 version of decorators, which is the
+    // "legacy" version. When decorators are used in legacy mode,
+    // proposal-class-properties must be used in loose mode
+    // see https://babeljs.io/docs/en/babel-plugin-proposal-decorators#note-compatibility-with-babel-plugin-proposal-class-properties
+    plugins.unshift(
+      [require.resolve('@babel/plugin-proposal-decorators'), {legacy: true}],
+    );
+
+    plugins.push(
+      require.resolve('@babel/plugin-proposal-numeric-separator'),
```
Comment (lemonmade) [inline non-standard-plugins.js:23]: Not automatically done by preset-env?

File: non-standard-plugins.js [hunk 3]
```diff
@@ -12,5 +12,21 @@ module.exports = function shopifyNonStandardPlugins(options = {}) {
...
 
+  if (typescript) {
+    // proposal-decorators must go before proposal-class-properties, so add this
+    // to the front of the array.
+    // Typscript implements the stage 1 version of decorators, which is the
+    // "legacy" version. When decorators are used in legacy mode,
+    // proposal-class-properties must be used in loose mode
+    // see https://babeljs.io/docs/en/babel-plugin-proposal-decorators#note-compatibility-with-babel-plugin-proposal-class-properties
+    plugins.unshift(
+      [require.resolve('@babel/plugin-proposal-decorators'), {legacy: true}],
```
Comment (lemonmade) [inline non-standard-plugins.js]: This doesn't need the syntax plugin too?

**Review by lemonmade — APPROVED** [review]
🥇

File: non-standard-plugins.js (hunk 2 above)
Comment (BPScott) [reply non-standard-plugins.js:23]: From what I can tell preset-env only contains stage 4 proposals. numeric-separator is still at stage 3.

File: non-standard-plugins.js (hunk 3 above)
Comment (BPScott) [reply non-standard-plugins.js]: It wasn't needed when they were defined in https://github.com/Shopify/sewing-kit/blob/15083d441339b588518aa1649581f4f509b8b56f/packages/sewing-kit/jest/transformers/typescript.js, so I assume not

File: non-standard-plugins.js (hunk 2 above)
Comment (lemonmade) [reply non-standard-plugins.js:23]: 👍 Can we maybe add it always, even without TS? Seems like a safe bet to include automatically for everyone.

File: non-standard-plugins.js (hunk 1 above)
Comment (BPScott) [reply non-standard-plugins.js]: I was initially resisting the need to add an `else` but it makes other stuff more readable 👍

File: non-standard-plugins.js (hunk 2 above)
Comment (BPScott) [reply non-standard-plugins.js:23]: I think think I'd prefer to keep it just for TS for now. We've been pretty good at not jumping on the experimental JS stuff bandwagon and I don't think this is worth breaking that streak. Give it time and it'll solve itself anyway as once it hits stage 4 then it'll be a part of preset-env.

**Review by GoodForOneFare — APPROVED** [review]

**Pull request merged**